### PR TITLE
Investigate some design tweaks over Crypto Implementation

### PR DESCRIPTION
**Pull request by Farhad-Shabani**
As a reference for suggestion made in #1238

* [ ] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Added entry in `.changelog/`

### CODE REVIEW

**Comment by Wizdave97** [comment]
This is a much better implementation for substrate chains

File: tendermint/src/merkle.rs [hunk 1]
```diff
@@ -61,88 +61,31 @@ pub trait MerkleHash {
...
-        Digest::update(self, right);
-
-        // Finalize the digest, reset the hasher state
-        let digest = self.finalize_reset();
-
-        copy_to_hash(digest)
+    /// Compute a simple Merkle root from vectors of arbitrary byte vectors.
+    /// The leaves of the tree are the bytes of the given byte vectors in
+    /// the given order.
+    fn simple_hash_from_byte_vectors(&mut self, byte_vecs: &[Vec<u8>]) -> Hash {
```
Comment (mzabaluev) [inline tendermint/src/merkle.rs:68]: Why does this need to be converted to a method?
With the freestanding function, we can't misuse the API by e.g. calling it on a digest object with non-reset digest state.
`simple_hash_from_byte_vectors` is a one-stop-shop function that creates the digest state internally, hashes the input and returns the result.

File: tendermint/src/merkle.rs [hunk 2]
```diff
@@ -61,88 +61,31 @@ pub trait MerkleHash {
...
+        self.hash_byte_vectors(byte_vecs)
     }
 }
 
 /// A wrapper for platform-provided host functions which can't do incremental
 /// hashing. One unfortunate example of such platform is Polkadot.
-pub struct NonIncremental<H>(PhantomData<H>);
-
-impl<H> Default for NonIncremental<H> {
+pub struct DefaultMerkleHash;
```
Comment (mzabaluev) [inline tendermint/src/merkle.rs:75]: I think this is wrong, we specifically don't want a stateless function to be the basis for default hashing implementation, when in most programming environments we can use the `sha2` crate.

File: tendermint/src/merkle.rs [hunk 3]
```diff
@@ -1,49 +1,49 @@
...
-pub trait MerkleHash {
+pub trait MerkleHash: DigestSha256 {
     // tmhash({})
     // Pre and post-conditions: the hasher is in the reset state
     // before and after calling this function.
-    fn empty_hash(&mut self) -> Hash;
+    fn empty_hash(&mut self) -> Hash {
+        // Get the output of an empty digest state.
+        Self::digest_sha256(&Vec::with_capacity(0))
+    }
```
Comment (mzabaluev) [inline tendermint/src/merkle.rs:24]: Why bake the implementation into the trait?
This pattern is used elsewhere, for example, in `VerificationPredicates`, but I don't think it's needed here.

File: tendermint/src/merkle.rs [hunk 4]
```diff
@@ -154,7 +97,7 @@ mod tests {
         let empty_tree_root = &hex::decode(empty_tree_root_hex).unwrap();
         let empty_tree: Vec<Vec<u8>> = vec![vec![]; 0];
 
-        let root = simple_hash_from_byte_vectors::<Sha256>(&empty_tree);
+        let root = DefaultMerkleHash.simple_hash_from_byte_vectors(&empty_tree);
```
Comment (mzabaluev) [inline tendermint/src/merkle.rs:100]: This looks non-idiomatic, the object instance has no purpose here, as explained above.

File: tendermint/src/merkle.rs [hunk 5]
```diff
@@ -61,88 +61,31 @@ pub trait MerkleHash {
...
-    fn empty_hash(&mut self) -> Hash {
-        // Get the output of an empty digest state.
-        let digest = self.finalize_reset();
-        copy_to_hash(digest)
-    }
-
-    fn leaf_hash(&mut self, bytes: &[u8]) -> Hash {
-        // Feed the data to the hasher, prepended with 0x00.
-        Digest::update(self, [0x00]);
-        Digest::update(self, bytes);
```
Comment (mzabaluev) [inline tendermint/src/merkle.rs:86]: Note how this does not allocate any memory temporarily or copy data around. This is what we want most of our users to actually use, while the few platforms where the `NonIncremental` wrapper is needed could be kept as integration freak cases.

**Review by mzabaluev — COMMENTED** [review]
Some initial comments, mostly flak. I think we'll need to come up with a minimal set of changes that make sense and reduce coupling between implementations and specific hash functions.

File: tendermint/Cargo.toml [hunk 6]
```diff
@@ -50,16 +50,15 @@ tendermint-proto = { version = "0.28.0", default-features = false, path = "../pr
 time = { version = "0.3", default-features = false, features = ["macros", "parsing"] }
 zeroize = { version = "1.1", default-features = false, features = ["zeroize_derive", "alloc"] }
 flex-error = { version = "0.4.4", default-features = false }
-sha2 = { version = "0.10", optional = true, default-features = false }
+sha2 = { version = "0.10", default-features = false }
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply tendermint/Cargo.toml:53]: Yeah, comparatively, it doesn't buy much. I looked it as a side objective

File: tendermint/src/merkle.rs [hunk 7]
```diff
@@ -1,49 +1,49 @@
...
-where
-    H: MerkleHash + Default,
-{
-    let mut hasher = H::default();
-    hasher.hash_byte_vectors(byte_vecs)
-}
-
 /// Implementation of Merkle tree hashing for Tendermint.
-pub trait MerkleHash {
+pub trait MerkleHash: DigestSha256 {
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply tendermint/src/merkle.rs:17]: It already do so. With the `DigestSha256` trait (included types from higher-level `HashType` trait), basically you can introduce any type of hashing functions. but I caught the issue. It should be renamed to make more sense (e.g.as DigestHash)

File: tendermint/src/merkle.rs [hunk 8]
```diff
@@ -1,49 +1,49 @@
...
+    }
 
     // tmhash(0x00 || leaf)
     // Pre and post-conditions: the hasher is in the reset state
     // before and after calling this function.
-    fn leaf_hash(&mut self, bytes: &[u8]) -> Hash;
+    fn leaf_hash(&mut self, bytes: &[u8]) -> Hash {
+        let mut buf = Vec::with_capacity(1 + bytes.len());
+        buf.push(0x00);
+        buf.extend_from_slice(bytes);
```
[the remark this answers is not in the document]
Comment (Farhad-Shabani) [reply tendermint/src/merkle.rs:32]: Did so, looking for a minimal set of changes compared to the orig files and hoping this be covered in a separate PR for this purpose. but Right, if there is a demand and a performance matter, would be better to get it back and also have it here.

File: tendermint/src/merkle.rs (hunk 1 above)
Comment (Farhad-Shabani) [reply tendermint/src/merkle.rs:68]: Likewise, I had looked at it from the opposite angle, but with the same concern. Having a bare function, so Merkle hasher could be triggered by a fake similar one. I encapsulated it to operate only on that data. btw, could be investigate more or even leave it untouched for now. alright

File: tendermint/src/merkle.rs (hunk 2 above)
Comment (Farhad-Shabani) [reply tendermint/src/merkle.rs:75]: I don’t know if I catch you correctly, but as it stands here, it is bounded by `sha2::Sha256`

**Comment by Farhad-Shabani** [comment]
In light of the need Mikhail aptly raised for incremental hashing and the comments mostly arose because of losing it, I've reverted that part. 
Here, just remains the suggestions you may want to consider, including the refactored parts for verifying signatures verification through an interface.

**Comment by tony-iqlusion** [comment]
@Farhad-Shabani I don't understand what advantage that signature verification interface is supposed to provide over [`signature::Verifier`](https://docs.rs/signature/latest/signature/trait.Verifier.html) and similar other signature verification traits in the `signature` crate.

It seems needlessly convoluted, and also has the disadvantage that it isn't natively supported by commonly used Rust ecosystem cryptography crates.

**Comment by Farhad-Shabani** [comment]
@tony-iqlusion agree with you on this. Actually, I've tried it first, but having `PublicKey` as enum, supporting different key/signature variants made it tricky to refactor and provide the needed interface for `validator/Info`. 
But, thanks bringing this up, made me rethought and came up with a new idea. Check that please. Now the `validator/Info`  can provide the generic interface through `verify_with` method. if it works overall, it can be fastened

**Pull request closed without merging**